lms/djangoapps/program_enrollments/api/writing.py

```diff
@@ -0,0 +1,426 @@
+"""
+Python API functions related to writing program enrollments.
+
+Outside of this subpackage, import these functions
+from `lms.djangoapps.program_enrollments.api`.
+"""
+from __future__ import absolute_import, unicode_literals
+
+import logging
+
+from course_modes.models import CourseMode
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from student.models import CourseEnrollment, NonExistentCourseError
+
+from ..constants import ProgramCourseEnrollmentStatuses
+from ..constants import ProgramCourseOperationStatuses as ProgramCourseOpStatuses
+from ..constants import ProgramEnrollmentStatuses
+from ..constants import ProgramOperationStatuses as ProgramOpStatuses
+from ..exceptions import ProviderDoesNotExistException
+from ..models import ProgramCourseEnrollment, ProgramEnrollment
+from .reading import fetch_program_course_enrollments, fetch_program_enrollments, get_users_by_external_keys
+
+logger = logging.getLogger(__name__)
+
+
+def write_program_enrollments(program_uuid, enrollment_requests, create, update):
```

Comment:
Why use the flags here in the single function? Why not keep the separate functions of `create` and `update`. But organize the common elements of those two into a single function? Maybe a context manager? Or even class hierarchy?  Same questions apply to the WriteProgramCourseEnrollments function.

Reply:
I thought about a couple different approaches before setting on this, and decided to settle on what I have in interest of time. It's definitely long for a single function, so I'm happy to spend a little longer trying to split this up better while still sharing functionality.

Reply:
We talked offline and decided that refactoring the function further is not a priority.

I agree that it would be good future tech-debt-bashing work.